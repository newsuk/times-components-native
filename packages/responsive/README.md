# Responsive

This package includes utilities for supporting responsive layouts for native
platforms. It provides a wrapper component called <Responsive> which listens to
orientation changes and provides layout information through context named
`ResponsiveContext`. Any child component can then use this layout information by
consuming `ResponsiveContext`.
