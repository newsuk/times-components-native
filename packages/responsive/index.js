export { default, ResponsiveContext } from './src/responsive';
