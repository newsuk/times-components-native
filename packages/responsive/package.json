{
  "name": "@times-components-native/responsive",
  "private": true,
  "version": "0.6.15",
  "description": "Responsive",
  "main": "src/responsive",
  "scripts": {
    "depcheck": "depcheck --ignores='@babel/*,babel-*,depcheck,eslint,jest,prettier' --ignore-bin-package=false --skip-missing",
    "fmt": "eslint . --fix && prettier --write '**/*.*'",
    "prettier:diff": "prettier --list-different '**/*.*'",
    "lint": "eslint . && yarn prettier:diff && yarn depcheck",
    "test:android": "jest --config='./__tests__/android/jest.config.js'",
    "test:ios": "jest --config='./__tests__/ios/jest.config.js'",
    "test:all": "yarn test:android --coverage && yarn test:ios --coverage"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/newsuk/times-components-native.git"
  },
  "keywords": [
    "react",
    "native",
    "responsive",
    "component"
  ],
  "license": "BSD-3-Clause",
  "bugs": {
    "url": "https://github.com/newsuk/times-components/issues"
  },
  "homepage": "https://github.com/newsuk/times-components#readme",
  "dependencies": {


    "prop-types": "15.7.2"
  },
  "resolutions": {
    "react": "16.9.0",
    "react-dom": "16.9.0"
  },
  "peerDependencies": {
    "react": ">=16.9",
    "react-dom": ">=16.9",
    "react-native": ">=0.59"
  },
  "publishConfig": {
    "access": "public"
  }
}
