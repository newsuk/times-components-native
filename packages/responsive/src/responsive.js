import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  getEditionBreakpoint,
  getNarrowArticleBreakpoint,
  tabletWidth,
} from "@times-components-native/styleguide";
import {
  getDimensions,
  addDimensionsListener,
  removeDimensionsListener,
} from "@times-components-native/utils";
import { NativeModules } from "react-native";
import ResponsiveContext from "./context";

const config = (NativeModules || {}).ReactConfig;

const calculateState = (width, fontScale) => ({
  editionBreakpoint: getEditionBreakpoint(width),
  narrowArticleBreakpoint: getNarrowArticleBreakpoint(width),
  fontScale,
  isTablet:
    (config && config.breakpoint && config.breakpoint !== "small") ||
    width >= tabletWidth,
  screenWidth: width,
});

class Responsive extends Component {
  constructor(props) {
    super(props);
    this.onDimensionChange = this.onDimensionChange.bind(this);
    const { fontScale, width } = getDimensions();
    this.state = calculateState(width, fontScale);
  }

  componentDidMount() {
    addDimensionsListener("change", this.onDimensionChange);
  }

  componentWillUnmount() {
    removeDimensionsListener("change", this.onDimensionChange);
  }

  onDimensionChange({ window: { fontScale, width } }) {
    const { fontScale: oldScale, screenWidth: oldWidth } = this.state;
    if (fontScale !== oldScale || (oldWidth && width !== oldWidth)) {
      console.log("UPDATE", fontScale, oldScale, width, oldWidth);
      this.setState(calculateState(width, fontScale));
    }
  }

  render() {
    const { children } = this.props;

    return (
      <ResponsiveContext.Provider value={this.state}>
        {children}
      </ResponsiveContext.Provider>
    );
  }
}

Responsive.propTypes = {
  children: PropTypes.node,
};

Responsive.defaultProps = {
  children: null,
};

export default Responsive;
export { ResponsiveContext };
