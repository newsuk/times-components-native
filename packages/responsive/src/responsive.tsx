import React, { useContext, useEffect, useState } from "react";
import {
  getDimensions,
  addDimensionsListener,
  removeDimensionsListener,
} from "@times-components-native/utils";
import { ScaledSize } from "react-native";
import ResponsiveContext from "./context";
import { calculateResponsiveContext } from "./calculateResponsiveContext";

interface DimensionChangeEvent {
  window: ScaledSize;
}

interface Props {
  initialHeight?: number | undefined;
  initialWidth?: number | undefined;
  initialFontScale?: number | undefined;
}

const ResponsiveProvider: React.FC<Props> = ({
  children,
  initialWidth,
  initialHeight,
  initialFontScale,
}) => {
  const { fontScale, width, height } =
    initialWidth && initialHeight && initialFontScale
      ? {
          width: initialWidth,
          height: initialHeight,
          fontScale: initialFontScale,
        }
      : getDimensions();

  const [state, setState] = useState(
    calculateResponsiveContext(width, height, fontScale),
  );

  const onDimensionChange = ({
    window: { fontScale, width, height },
  }: DimensionChangeEvent) =>
    setState(calculateResponsiveContext(width, height, fontScale));

  useEffect(() => {
    const listener = addDimensionsListener("change", onDimensionChange);

    return () => {
      removeDimensionsListener("change", listener);
    };
  }, []);

  return (
    <ResponsiveContext.Provider value={state}>
      {children}
    </ResponsiveContext.Provider>
  );
};

const useResponsiveContext = () => useContext(ResponsiveContext);

export { ResponsiveContext, useResponsiveContext };
export default ResponsiveProvider;
