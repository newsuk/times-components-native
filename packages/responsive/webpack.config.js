/* eslint-disable import/no-extraneous-dependencies */
module.exports = require("@times-components-native/webpack-configurator")(
  __dirname,
  "dev"
);
