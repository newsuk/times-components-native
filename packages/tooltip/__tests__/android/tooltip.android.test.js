import Tooltip from "../../../tooltip";
import TestRenderer from "react-test-renderer";
import { ResponsiveContext } from "@times-components-native/responsive";
import { shallow } from "enzyme";
import { TouchableOpacity } from "react-native";
import React from "react";
import { Text } from "react-native";
import "../serializers-with-all-styles";
import { Platform } from "react-native";

export const withTabletContext = (WrappedComponent, isTablet = true) => (
  <ResponsiveContext.Provider
    value={{
      isTablet: isTablet,
    }}
  >
    {WrappedComponent}
  </ResponsiveContext.Provider>
);

describe("Tooltip", () => {
  Platform.OS = "android";
  it("hides tooltip close button on android only", () => {
    Platform.OS = "android";

    const onTooltipPresentedMock = jest.fn();

    const output = TestRenderer.create(
      <Tooltip
        content={<Text>foo</Text>}
        onTooltipPresented={onTooltipPresentedMock}
        type="testtype"
        tooltips={["testtype"]}
        placement="top"
      >
        bar
      </Tooltip>,
    );
    expect(output).toMatchSnapshot();
  });
  it("onClose is called correctly", async () => {
    const onTooltipPresentedMock = jest.fn();
    const onCloseMock = jest.fn();

    const testInstance = shallow(
      <Tooltip
        content={<Text>foo</Text>}
        onClose={onCloseMock}
        onTooltipPresented={onTooltipPresentedMock}
        type="testtype"
        tooltips={["testtype"]}
      >
        bar
      </Tooltip>,
    );

    const closeButton = testInstance.find(TouchableOpacity);

    expect(closeButton.exists()).toBeFalsy();
  });
});
