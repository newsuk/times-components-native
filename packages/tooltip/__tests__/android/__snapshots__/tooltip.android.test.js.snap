// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip hides tooltip close button on android only 1`] = `
<View
  style={
    Object {
      "zIndex": 9999,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#006699",
            "borderRadius": 3,
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 3,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 3,
            "width": 256,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "GillSansMTStd-Medium",
              "fontSize": 16,
              "lineHeight": 18,
              "paddingHorizontal": 30,
              "paddingRight": 35,
              "paddingVertical": 20,
              "textAlign": "left",
            }
          }
        >
          <Text>
            foo
          </Text>
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "transparent",
              "borderLeftColor": "transparent",
              "borderLeftWidth": 12,
              "borderRightColor": "transparent",
              "borderRightWidth": 12,
              "borderStyle": "solid",
              "borderTopColor": "#006699",
              "borderTopWidth": 8,
              "bottom": -8,
              "left": 20,
              "position": "absolute",
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 3,
                "width": 3,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4,
              "top": "auto",
              "transform": Array [
                Object {
                  "rotateX": "0deg",
                },
              ],
            }
          }
        />
      </View>
    </View>
  </View>
  bar
</View>
`;
